core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -113,23 +111,46 @@ protected boolean resolveIndex() {
     @Override
     public WritePrimaryResult<BulkShardRequest, BulkShardResponse> shardOperationOnPrimary(
             BulkShardRequest request, IndexShard primary) throws Exception {
+        final BulkShardResult shardResult = performOnPrimary(request, primary, updateHelper,
+                threadPool::absoluteTimeInMillis, new ConcreteMappingUpdatePerformer());
+        return new WritePrimaryResult<>(shardResult.request, shardResult.response,
+                shardResult.location, null, primary, logger);
+    }
+
+    /** Result holder of executing shard bulk on primary */
+    public static class BulkShardResult {
```

Comment:
I initially used `WritePrimaryResult` for this, but that changed the visibility of `WritePrimaryResult` from protected to public. Hence I introduced `BulkShardResult` which allows us to use `TransportShardBulkAction#performOnPrimary` in `ESIndexLevelReplicationTestCase` without exposing `WritePrimaryResult`